File: SRM 718/InterleavingParenthesis.java
import java.io.*;
import java.util.*;

public class InterleavingParenthesis {
    private static int MOD = 1000_000_007;
    int[][] dp;
	public int countWays(String s1, String s2) {
        int n = s1.length();
        int m = s2.length();
        dp = new int[s1.length() + 1][s2.length() + 1];
        int[] a = new int[s1.length() + 1];
        int[] b = new int[s2.length() + 1];
        for (int i = 1; i < s1.length() + 1; ++i) {
            if (s1.charAt(i - 1) == '(') a[i] = a[i - 1] + 1;
            else a[i] = a[i - 1] - 1;
        }
        for (int i = 1; i < s2.length() + 1; ++i) {
            if (s2.charAt(i - 1) == '(') b[i] = b[i - 1] + 1;
            else b[i] = b[i - 1] - 1;
        }
        dp[0][0] = 1;
        for (int i = 0; i < s1.length() + 1; ++i) {
            for (int j = 0; j < s2.length() + 1; ++j) {
                if (i == 0 && j == 0) {
                    continue;
                }
                dp[i][j] = 0;
                if (a[i] + b[j] < 0) continue;
                if (i > 0) dp[i][j] = dp[i-1][j];
                if (j > 0) dp[i][j] = (dp[i][j] + dp[i][j-1]) % MOD;
            }
        }
		return dp[n][m];
	}

// CUT begin
	public static void main(String[] args){
		System.err.println("InterleavingParenthesis (250 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            String s1 = Reader.nextLine();
            String s2 = Reader.nextLine();
            Reader.nextLine();
            int __answer = Integer.parseInt(Reader.nextLine());

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(s1, s2, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1500105392;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 250 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(String s1, String s2, int __expected) {
        s1 = new String(s1);
        s2 = new String(s2);
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		InterleavingParenthesis instance = new InterleavingParenthesis();
		int __result = 0;
		try {
			__result = instance.countWays(s1, s2);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "InterleavingParenthesis.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}

File: SRM 718/InterleavingParenthesis.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>

using namespace std;
#define CLR(x) memset(x, 0, sizeof(x))
#define FOR(i,a,b) for(int i = (a); i < (b); ++i)
#define FORE(i,a,b) for(int i = (a); i <= (b); ++i)
const int MX = 3000;
const int MOD = 1e9 + 7;
int a[MX], b[MX], dp[MX][MX];
//int n,m;
class InterleavingParenthesis {
    public:
    int countWays(string s1, string s2) {
        int n = s1.length();
        int m = s2.length();
        CLR(a); CLR(b); CLR(dp);
        FORE(i, 1, n) {
            if (s1[i - 1] == '(') {
                a[i] = a[i - 1] + 1;
            } else {
                a[i] = a[i - 1] - 1;
            }
        }
        FORE(i, 1, m) {
            if (s2[i - 1] == '(') {
                b[i] = b[i - 1] + 1;
            } else {
                b[i] = b[i - 1] - 1;
            }
        }
        dp[0][0] = 1;
        FORE(i, 0, n) FORE(j, 0, m) {
            if (i == 0 && j == 0) {
                continue;
            }
            dp[i][j] = 0;
            if (a[i] + b[j] < 0) {
                continue;
            }
            if (i > 0) {
                dp[i][j] = dp[i-1][j];
            }
            if (j > 0) {
                dp[i][j] = (dp[i][j] + dp[i][j-1]) % MOD;
            }
        }
        if (a[n] + b[m] != 0) return 0;
        return dp[n][m];
    }
};

// CUT begin
ifstream data("InterleavingParenthesis.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(string s1, string s2, int __expected) {
    time_t startClock = clock();
    InterleavingParenthesis *instance = new InterleavingParenthesis();
    int __result = instance->countWays(s1, s2);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        string s1;
        from_stream(s1);
        string s2;
        from_stream(s2);
        next_line();
        int __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(s1, s2, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1500108923;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 250 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "InterleavingParenthesis (250 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end
